File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn is-valid? [a b c]
  (and (every? #(> % 0) [a b c])
       (>= (+ a b) c)
       (>= (+ a c) b)
       (>= (+ b c) a)))

(defn equilateral? [a b c]
  (and (is-valid? a b c)
       (= a b c)))

(defn isosceles? [a b c]
  (and (is-valid? a b c)
       (or (= a b)
           (= a c)
           (= b c))))

(defn scalene? [a b c] 
  (and (is-valid? a b c)
       (= 3 (count (hash-set a b c)))))

File: ruby/triangle/triangle.rb
class Triangle
  def initialize(triangle)
    @a = triangle[0]
    @b = triangle[1]
    @c = triangle[2]
  end

  def valid?
    [@a, @b, @c].all?{ |el| el > 0 } &&
    @a + @b > @c &&
    @a + @c > @b &&
    @b + @c > @a
  end

  def equilateral?
    valid? && [@a, @b, @c].uniq.length == 1
  end

  def isosceles?
    valid? && (
      @a == @b ||
      @a == @c ||
      @b == @c
    )
  end

  def scalene?
    valid? && (
      @a != @b &&
      @a != @c &&
      @b != @c
    )
  end
end
